programs/dynamic-bonding-curve/src/state/virtual_pool.rs

```diff
@@ -276,101 +426,290 @@ impl VirtualPool {
                 continue;
             }
             if config.curve[i].sqrt_price < current_sqrt_price {
-                let max_amount_in = get_delta_amount_base_unsigned_256(
+                let max_amount_out = get_delta_amount_quote_unsigned_256(
                     config.curve[i].sqrt_price,
                     current_sqrt_price,
                     config.curve[i + 1].liquidity,
-                    Rounding::Up, // TODO check whether we should use round down or round up
+                    Rounding::Down,
                 )?;
-                if U256::from(amount_left) < max_amount_in {
-                    let next_sqrt_price = get_next_sqrt_price_from_input(
+                if U256::from(amount_left) < max_amount_out {
+                    let next_sqrt_price = get_next_sqrt_price_from_output(
                         current_sqrt_price,
                         config.curve[i + 1].liquidity,
                         amount_left,
                         true,
                     )?;
 
-                    let output_amount = get_delta_amount_quote_unsigned(
+                    let input_amount = get_delta_amount_base_unsigned(
                         next_sqrt_price,
                         current_sqrt_price,
                         config.curve[i + 1].liquidity,
-                        Rounding::Down,
+                        Rounding::Up,
                     )?;
-                    total_output_amount = total_output_amount.safe_add(output_amount)?;
+
+                    total_input_amount = total_input_amount.safe_add(input_amount)?;
                     current_sqrt_price = next_sqrt_price;
                     amount_left = 0;
                     break;
                 } else {
                     let next_sqrt_price = config.curve[i].sqrt_price;
-                    let output_amount = get_delta_amount_quote_unsigned(
+                    let input_amount = get_delta_amount_base_unsigned(
                         next_sqrt_price,
                         current_sqrt_price,
                         config.curve[i + 1].liquidity,
-                        Rounding::Down,
+                        Rounding::Up,
                     )?;
-                    total_output_amount = total_output_amount.safe_add(output_amount)?;
+                    total_input_amount = total_input_amount.safe_add(input_amount)?;
                     current_sqrt_price = next_sqrt_price;
                     amount_left = amount_left.safe_sub(
-                        max_amount_in
+                        max_amount_out
                             .try_into()
                             .map_err(|_| PoolError::TypeCastFailed)?,
                     )?;
                 }
             }
         }
         if amount_left != 0 {
-            let next_sqrt_price = get_next_sqrt_price_from_input(
+            let next_sqrt_price = get_next_sqrt_price_from_output(
```

Comment:
We need to validate `next_sqrt_price ` is greater than or equal config.sqrt_start_price

Reply:
Fixed